### PR TITLE
Add Intel gpu dialect

### PR DESCRIPTION
**Pull request by nbpatel**
Please review these guidelines to help with the review process:
- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?
- [x] Have you organized your commits logically and ensured each can be built by itself?

This PR adds the code for Intel Gpu Dialect. Its a dialect with two ops. The RFC is in #323. will add simple test case.

### CODE REVIEW

File: include/imex/Dialect/IntelGpu/IR/IntelGpuOps.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <mlir/IR/BuiltinTypes.h>
+#include <mlir/IR/Dialect.h>
+#include <mlir/IR/OpDefinition.h>
+#include <mlir/IR/Types.h>
+#include <mlir/Interfaces/SideEffectInterfaces.h>
+
+namespace imex {
+namespace intel_gpu {
+class OpaqueType : public ::mlir::Type::TypeBase<OpaqueType, ::mlir::Type,
```
Comment (fschlimb) [inline include/imex/Dialect/IntelGpu/IR/IntelGpuOps.h:26]: Why is this not in the .td file?

File: include/imex/Dialect/IntelGpu/IR/IntelGpuOps.td [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+//   * The parent dialect of the operation.
+//   * The mnemonic for the operation, or the name without the dialect prefix.
+//   * A list of traits for the operation.
+class IntelGpu_Op<string mnemonic, list<Trait> traits = []> :
+    Op<IntelGpu_Dialect, mnemonic, traits>;
+
+def IntelGpu_OpaqueType
+    : DialectType<IntelGpu_Dialect, CPred<"$_self.isa<::imex::intel_gpu::OpaqueType>()">,
+                  "opaque_type">,
+      BuildableType<"$_builder.getType<::imex::intel_gpu::OpaqueType>()"> {}
```
Comment (fschlimb) [inline include/imex/Dialect/IntelGpu/IR/IntelGpuOps.td:54]: What is this good for? It has no summary and no description.

File: include/imex/Dialect/IntelGpu/IR/IntelGpuOps.td [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+def CreateStreamOp : IntelGpu_Op<"create_stream"> {
+
+  // result is a Stream
+  let results = (outs IntelGpu_OpaqueType : $stream);
+}
```
Comment (fschlimb) [inline include/imex/Dialect/IntelGpu/IR/IntelGpuOps.td:60]: Please add summary and description for each an every op/type.

File: lib/Dialect/IntelGpu/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+add_subdirectory(IR)
```
Comment (mshahneo) [inline lib/Dialect/IntelGpu/CMakeLists.txt]: @nbpatel, just wanted to point this out. Maybe clang-format issue. Might be better to use the clang-format to format.

**Pull request closed without merging**